django/db/models/sql/query.py

```diff
@@ -1401,10 +1403,50 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
         key field for example).
         """
         joins = [alias]
-        # First, generate the path for the names
-        path, final_field, targets, rest = self.names_to_path(
-            names, opts, allow_many, fail_on_missing=True)
-
+        # We can't apply the transforms yet, as we still need to trim joins
+        # later in the process. However, we don't want to make every user
+        # of this function look up transforms directly, so we compute them
+        # here and create a partial that will convert fields to the appropriate
+        # wrapped version.
+
+        def final_transformer(field, alias):
+            return field.get_col(alias)
+
+        # Try resolving all the names as fields first, but if there is an error
+        # progressively treat trailing names as lookups until we can resolve
+        # a field
+        last_field_exception = None
+        for pivot in range(len(names), 0, -1):
+            try:
+                path, final_field, targets, rest = self.names_to_path(
+                    names[:pivot], opts, allow_many, fail_on_missing=True)
+            except FieldError as exc:
+                if pivot == 1:
+                    # The first item cannot be a lookup, so it's safe
+                    # to raise the field error here.
+                    raise
+                else:
+                    last_field_exception = exc
+            else:
+                # The transforms are the remaining items that couldn't be
+                # resolved into fields
+                transforms = names[pivot:]
+                break
+        for name in transforms:
+            def transform(field, alias, name, previous):
```

Comment:
`def transform(field, alias, *, name, previous)` ?

Makes the partial application further down more obvious that name and previous are supplied by keyword args. Not certain this is better, but I had to double check the partial applicability worked correctly.

Reply:
Agreed.